lib/src/rules/prefer_final_locals.dart

```diff
@@ -76,6 +81,101 @@ class _Visitor extends SimpleAstVisitor<void> {
 
   _Visitor(this.rule);
 
+  void checkPatternFields(DartPattern node) {
+    var function = node.thisOrAncestorOfType<FunctionBody>();
+    if (function == null) return;
+
+    late NodeList<PatternField> fields;
+    if (node is RecordPattern) fields = node.fields;
+    if (node is ObjectPattern) fields = node.fields;
+
+    var parent = node.unParenthesized.parent;
+    var inPatternVariableDeclaration = false;
+    if (parent is PatternVariableDeclaration) {
+      if (parent.keyword.keyword == Keyword.FINAL) return;
+      inPatternVariableDeclaration = true;
+    }
+
+    for (var field in fields) {
+      var pattern = field.pattern.declaredVariablePattern;
+      if (pattern is DeclaredVariablePattern) {
+        var element = pattern.declaredElement;
+        if (element == null) continue;
+        if (function.isPotentiallyMutatedInScope(element)) {
+          if (inPatternVariableDeclaration) {
+            return;
+          } else {
+            continue;
+          }
+        }
+        if (inPatternVariableDeclaration) {
+          rule.reportLint((parent! as PatternVariableDeclaration).expression);
+          return;
+        } else {
+          if (!pattern.keyword.isFinal) {
+            rule.reportLintForToken(pattern.name);
+          }
+        }
+      }
+    }
+  }
+
+  bool isDeclaredFinal(AstNode node) {
+    var declaration = node.thisOrAncestorOfType<PatternVariableDeclaration>();
+    if (declaration == null) return false; // To be safe.
+    return declaration.keyword.isFinal;
+  }
+
+  bool isPotentiallyMutated(AstNode pattern, FunctionBody function) {
```

Comment:
This is attempting to return `true` if any variable in the `pattern` is potentially mutated, but it doesn't appear to account for variables nested inside other patterns (such as a logical-and or list pattern).

Reply:
Addressed in https://github.com/dart-lang/linter/pull/4225